### PR TITLE
Infra, WPT, CI

### PR DESCRIPTION
**Pull request by zouhir**
# Bundling Source Files to a Synchronous IIFE Module

### Proposed solution
Compile the es-modules JavaScript source files with [microbundle](https://github.com/developit/microbundle) and produce an iife (Immediately invoked function expression) that synchronously appends ScrollTimeline constructor to the Window object.

### Why microbundle as a bundler of choice
This bundler module is capable of creating a small, optimized JS bundle from source files using rollup under the hood with Zero-configurations needed. 
It is currently used by some other project and polyfills such:

- [GoogleChromeLabs/css-paint-polyfill](https://github.com/GoogleChromeLabs/css-paint-polyfill/blob/master/package.json#L8)
- [GoogleChromeLabs/kv-storage-polyfill](https://github.com/GoogleChromeLabs/kv-storage-polyfill/blob/master/package.json#L9)
- [GoogleChromeLabs/audioworklet-polyfill](https://github.com/GoogleChromeLabs/audioworklet-polyfill/blob/master/package.json#L17)

# E2E Testing

Tests currently run against WPT in CI, it uses Saucelabs VMs to test in Edge, Chrome, FF.

Here's a screenshot of the current results:

Screen Shot 2020-06-12 at 9 09 32 PM

For comparison, it looks like this in a browser (not polyfill):

Screen Shot 2020-06-12 at 9 09 41 PM

## Why saucelabs?
Cypress.io has a nice API for writing automated browser test but it does not support Safari and support for FireFox is in Beta, that makes Saucelabs the best option for automated browser testing at this moment.

### 1: The Web-animations-js polyfill strategy
That polyfill have a mix of: JavaScript unit tests and e2e integration tests, all run in CI via Saucelabs which we are planning to use for ScrollTimeline. 

Unit tests written in Mocha and running against real browser APIs (not [JSDOM](https://github.com/jsdom/jsdom)) via [Karma test runner](https://karma-runner.github.io/latest/index.html).

I believe the e2e integration test setup (for WPT, Blink test files) is counter-intuitive and hard to maintain, for those reasons:
- Karma excels at running JS tests against DOM or web pages as a context, and not running tests in HTML files.
- Web-animations-js polyfills has a long chain of configurations: a [framework-specific task runner modules ](https://github.com/web-animations/web-animations-js/blob/dev/Gruntfile.js#L20) that pre-processes [target files](https://github.com/web-animations/web-animations-js/blob/dev/target-config.js) and has a modified testharnessreporter (forked from WPT) as well as custom Karma test harness runners you [can see here](https://github.com/web-animations/web-animations-js/tree/dev/test). Being as close as possible to WPT files is beneficial and the more pre-processing and customization, the harder it will be for us to maintain.

### 2: Philip Walton's Blog Post on using Saucelabs Rest API [read](https://philipwalton.com/articles/learning-how-to-set-up-automated-cross-browser-javascript-unit-testing/)

The solution outline in Philip Walton's (Engineer @ Google) Blog is more straight-forward than Web-animations-js polyfill solution, he uses Mocha (the client side library) to run tests inside the HTML pages, and post those pages to sauce labs using their REST API, all needed there is a custom "failLogger" when the tests are finished, which is 1 JS function that acts as an adapter between Mocha results and Saucelabs. 

One note is that the Saucelabs rest API does not support localhost:xxxx addresses, the developer in that post used [ngrok](https://ngrok.com/) and mentioned using Sauce Connect VPN tunnel if it's possible

Frame 1 (1)

### 3: (✅currently using) Using Webdriver with Saucelabs

I found connect WPT from lcoalhost:xxxx to Saucelabs via a sauce connect tunnel and using selenium web driver to test against a matrix of browsers is the simplest & most effective way to run WPT harness tests, and even do more if needed (screenshots, ref, etc).

Some of the pros:

- not much pre-processing needed: we only need to inject the polyfill script below the `WPT/**/*.html` test `</title>` 
- all tests are ran via WPT test runner, no customization to the runner or reporter.

### 4: WPT test runner --sauce flag

Monday, April 20, Stephen McGruer from the Chromium project highlighted on the slack channel that we can use --sauce-* flags, [they are documented here](https://web-platform-tests.org/running-tests/command-line-arguments.html?highlight=sauce) I believe if this works after we inject the HTML, will even be an improvements to saucelabs' Web Driver implementation

We are tracking progress of this method here: https://github.com/web-platform-tests/wpt/issues/23898

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,3 +21,68 @@ document.getElementById('parallax').animate(
           endScrollOffset: '200px'})
     });
 ```
+
+# Contributing
+
+### Before you start
+
+You can simply copy `.env-no-secret` and rename it as `.env` and add your Saucelab secrets. This file should contain the following
```
[the remark this answers is not in the document]
Comment (zouhir) [reply README.md]: I re-arranged the section, now should be clearer and the new readme structure should resolve this, please lmk

File: test/setup/wpt-repo.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+PROJ_DIR="$(cd "$(dirname "$1")" && pwd)"
+
+WPT_DIR="$PROJ_DIR/test/wpt"
+
+# if directory does not exist, clone fresh-stuff from WPT master
+if [ ! -d "$WPT_DIR" ]; then
+  echo "WPT folder not found, cloning from remote..."
+  git clone https://github.com/web-platform-tests/wpt $WPT_DIR
```
[the remark this answers is not in the document]
Comment (zouhir) [reply test/setup/wpt-repo.sh:10]: I have updated this to:

- if WPT folder does not exist: clone
- if WPT folder exists: pull latest and try to rebase

please lmk what you thoink

File: test/setup/wpt-repo.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+fi
+
+if [ ! -d "$WPT_DIR/polyfills" ]; then
+  mkdir "$WPT_DIR/polyfills"
+fi
+
+rm -rf "$WPT_DIR/polyfills/scroll-timeline.js"
+cp "$PROJ_DIR/dist/scroll-timeline.js" "$WPT_DIR/polyfills/"
```
[the remark this answers is not in the document]
Comment (zouhir) [reply test/setup/wpt-repo.sh]: done.

File: test/tests.config.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
+  "polyfillFiles": [
+    "/polyfills/scroll-timeline.js"
+  ],
+  "harnessTests": [
+    "/scroll-animations/current-time-nan.html",
```
[the remark this answers is not in the document]
Comment (zouhir) [reply test/tests.config.json]: done

**Comment by zouhir** [comment]
@flackr I addressed the feedback, however, we have an issue with FireFox in CI. I will investigate tomorrow

File: test/setup/wpt-repo.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (flackr) [reply test/setup/wpt-repo.sh:10]: Sounds good!

File: test/tests.config.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (flackr) [reply test/tests.config.json]: Looks great!

File: test/setup/wpt-repo.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (flackr) [reply test/setup/wpt-repo.sh]: Awesome! No more modification of wpt necessary :-)

**Pull request merged**